## Data Analytics II 

Q. Implement Logistic Regression using Python to perform classification on Social_Network_ads.csv dataset.

Q. Compute confusion matrix to find TP, FP, TN, FN, Accuracy, Error rate, Precision, Recall on the given dataset.

This is performed by **Warepam Richard Singh**.

Check him out: https://linktr.ee/richardwarepam16

## Importing the necessary libraries:

In [26]:
import numpy as np
import pandas as pd

## Loading the Dataset

In [4]:
df = pd.read_csv('/Users/macbookpro/Desktop/Data-Science-pct/Sr.No.-5/Social_Network_Ads.csv')
df.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

## Data Analysis:

**Use and analyse the data with info () and describe ():**

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 15.8+ KB


In [6]:
df.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

## EDA will be done when we start using Data Visualization

EDA means Exploratory Data Analysis

## Let's assign x and y with the data and the target to predict

In [9]:
x = df.iloc[:,[2,3]].values
y = df.iloc[:,4].values

## Splitting the dataset into a training set and a testing set.

In [10]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y,train_size= 0.25, random_state= 101)

## Calling Logistic Regression as an instance

In [13]:
from sklearn.linear_model import LogisticRegression
log = LogisticRegression()
log.fit(x_train,y_train)

LogisticRegression()

## PREDICTION:

In [14]:
pred = log.predict(x_test)

## Classification report and Confusion Matrix of the Prediction

In [21]:
from sklearn.metrics import classification_report, confusion_matrix
print(classification_report(y_test, pred ))
print('\n')
print(confusion_matrix(y_test , pred ))

              precision    recall  f1-score   support

           0       0.64      1.00      0.78       193
           1       0.00      0.00      0.00       107

    accuracy                           0.64       300
   macro avg       0.32      0.50      0.39       300
weighted avg       0.41      0.64      0.50       300



[[193   0]
 [107   0]]


/Users/macbookpro/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/macbookpro/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/macbookpro/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## NOTE!!

Somewhere while predicting we can't find the precision, recall and f1-score of 1; as it is ill defined in the dataset. We can fix this by scaling the data 

## Scaling the Dataset

In [22]:
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)

In [23]:
log.fit(x_train, y_train)

LogisticRegression()

## PREDICTION AFTER SCALING

In [24]:
sc_pred = log.predict(x_test)

## Classification Report and Confusion Matrix after Scaling

In [25]:
from sklearn.metrics import classification_report, confusion_matrix
print(classification_report(y_test, sc_pred))
print('\n')
print(confusion_matrix(y_test , sc_pred))

              precision    recall  f1-score   support

           0       0.82      0.93      0.87       193
           1       0.83      0.64      0.73       107

    accuracy                           0.83       300
   macro avg       0.83      0.79      0.80       300
weighted avg       0.83      0.83      0.82       300



[[179  14]
 [ 38  69]]


The accuarcy turns from **64%** to **83%** without even an error about the Precision , recall and f1-score.

## THANK YOU!